[JadeCheah] iP

JadeCheah opened the pull request:
# Bartholomew 
> "Mine eyes smell onions. " — William Shakespeare
>
Greetings, good traveler! I am Bartholomew, thy most faithful scribe. Entrust thy tasks to me, and I shall remember them for thy.  I am:
- Jolly 
- Nimble-witted
- ~Promptitude~ _Handy-dandy_ 
- _**FREE!!!**_

### How to summon me:
1. Go to this [link](https://github.com/JadeCheah/ip/releases/tag/A-Jar)
2. **Download the JAR file** and save it in a folder of your choice.
3. Open a command prompt (or terminal) and **navigate to the folder** where the JAR file is located:
    - On Windows: `cd path\to\your\folder`
  
    - On Mac/Linux: `cd /path/to/your/folder`
4. Run the following command to **start Bartholomew**:
    ```bash
    java -jar bartholomew.jar
    ```
5. Bartholomew is now ready to manage your tasks!

### Features:
- [ ] 📜 : **Keeps track of thy tasks** – So naught is forgotten.
- [x] ⌛ : **Manages deadlines** _(coming soon!)_ – No duty shall be left undone.
- [ ] 🔔 : **Sends reminders** _(coming soon!)_ – Like a faithful herald calling out thy tasks.
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JadeCheah committed:
Level-0 greet user done

JadeCheah committed:
Level-1 echo user done

JadeCheah committed:
Level-2 Add, List done

JadeCheah committed:
added Task.java file

JadeCheah committed:
added Level-3 Mark As done functionality

JadeCheah committed:
Level-4 todo, event, deadline done

JadeCheah committed:
Level-5 handle basic errors done

JadeCheah committed:
Level-6 delete task done

JadeCheah committed:
minor changes to enum taskType

JadeCheah committed:
A-TextUiTesting, basic ui testing done with color

JadeCheah committed:
Do level-7

JadeCheah committed:
Merge branch 'branch-Level-7' into master
Merge branch-Level-7  into master

This merge implements automatic saving and loading of tasks from disk (Level 7).
Tasks are now saved to ./data/duke.txt whenever the task list changes. When the
chatbot starts, it loads the tasks from this file if it exists.

Changes introduced:
- Implemented Storage class to handle file I/O operations.
- Tasks are stored in a structured format (e.g., `T | 1 | read book`).
- Ensured relative paths (`./data/duke.txt`) for cross-platform compatibility.
- Automatically creates the file and folder if they do not exist.
- Added error handling for missing or corrupted files.

With this change, users no longer lose their task list when the chatbot restarts.

JadeCheah committed:
do Level-8

JadeCheah committed:
Merge branch 'branch-Level-8' into master
Implemented date handling using LocalDate for Deadline and Event tasks.
Enhanced error handling for invalid date formats.

JadeCheah committed:
Add classes Command, Parser, Ui, and TaskList to increase OOP

JadeCheah committed:
Add more OOP to AddCommand but still have room for improvements

JadeCheah committed:
Merge branch 'A-MoreOOP' into master
Add new classes Command, TaskList, Parser, Ui, Storage to improve OOP.
Add new classes AddCommand, DeleteCommand, ListCommand, MarkCommand which extends abstract class Command
Move logic from Bartholomew class into respective classes.

JadeCheah committed:
Move classes into respective packages

JadeCheah committed:
Merge branch 'add-gradle-support' into master

JadeCheah committed:
Set up gradle and modify .gitignore file

JadeCheah committed:
Add JUnit, add TaskListTest and AddCommandTest

JadeCheah committed:
Merge branch 'A-JUnit' into master
Add JUnit, add TaskListTest and AddCommandTest

JadeCheah committed:
Change build.gradle file to create a JAR File

JadeCheah committed:
Add JavaDoc header comments to all classes and their methods, except Ui

JadeCheah committed:
Change switch indentation in AddCommand to fit coding standard

JadeCheah committed:
Implement find by keyword function

JadeCheah committed:
Merge branch 'branch-A-JavaDoc' into master

Add JavaDoc header comment to all classes and its methods except Ui

JadeCheah committed:
Merge branch 'branch-A-CodingStandard' into master

Tweak the code to comply with a given coding standard. Change switch case indentation in AddCommand.java.

JadeCheah committed:
Resolve merge conflicts

basil-boh commented on data/bart.txt:
Actually this file could be included in the .gitignore to avoid unnecessary version control

basil-boh commented on src/main/java/bart/Bartholomew.java:
Overall good abstractions and java coding conventions are generally adhered to, great job:)

basil-boh commented on src/main/java/bart/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+        return tasks.size();
+    }
+
+    public ArrayList<Task> findTask(String keyword) {
```
Nothing wrong with this! But can also consider using streams for filtering tasks which could be easier to maintain/modify in the future 

```
public ArrayList<Task> findTask(String keyword) {
    return tasks.stream()
        .filter(task -> task.getDescription().contains(keyword))
            .collect(Collectors.toCollection(ArrayList::new));
}
```

basil-boh commented on src/main/java/bart/command/AddCommand.java:
```diff
@@ -0,0 +1,129 @@
+            TaskType type = TaskType.valueOf(taskTypeStr);
+            switch (type) {
+            case TODO:
+                    if (taskDetails.isBlank()) {
```
I believe the indentation for the if statements should have 4 spacings

basil-boh commented on src/main/java/bart/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+ * Represents a list of tasks.
+ */
+public class TaskList {
+    public ArrayList<Task> tasks;
```
Might be careless but this breaks encapsulation and should be made private!

basil-boh commented on src/main/java/bart/util/Storage.java:
```diff
@@ -0,0 +1,138 @@
+
+            // Write tasks to the file
+            FileWriter writer = new FileWriter(filePath);
+            for (Task task : tasks.tasks) {
```
This also breaks encapsulation as tasks.tasks is accessing its class field. Should use a getter instead

basil-boh commented on src/main/java/bart/command/AddCommand.java:
```diff
@@ -0,0 +1,129 @@
+                    }
+                    newTask = new Todo(taskDetails.trim());
+                    break;
+            case DEADLINE:
```
Each of the cases seem to be handling too much logic. Can consider implementing methods for parsing each task, i.e parseToDoTask, parseDeadlineTask, etc

JadeCheah committed:
minor style changes

JadeCheah committed:
Add checkstyle.xml and suppressions.xml file, configurate checkstyle

JadeCheah committed:
Merge branch 'branck-A-CheckStyle' into master

JadeCheah committed:
Add JavaFX for Gui, To be integrated later

JadeCheah committed:
Integrate GUI into Bart fully.

- Delete most of Ui methods and change them to one that formats string to be returned.
- Create a CommandResult
- Modify execute method of Command Type to return a CommandResult Type
- Delete listTasks() from TaskList
- Fix style violations, write JavaDoc comments

JadeCheah committed:
Merge branch 'branch-Level-10' into master

JadeCheah committed:
Add greeting message upon starting

JadeCheah committed:
Delete text-ui-text folder

JadeCheah committed:
Clear bart.txt file

JadeCheah committed:
Add assertions to enforce key assumptions

Ensure critical assumptions hold, preventing invalid states.
Added assertions for:
* component initialization
* task access
* UI elements
* user input validation

JadeCheah committed:
Refactor UI setup and improve task index handling

Refactored UI initialization in Main.java:
- Extracted UI setup into a separate method `setupUI` for better readability and maintainability.
- Improved error handling by printing an error message when UI loading fails.

Enhanced task indexing in TaskList.java:
- Introduced `INDEX_OFFSET` constant to improve clarity and avoid magic numbers when adjusting for zero-based indexing.
- Updated `getTask` and `deleteTask` methods to use `INDEX_OFFSET`, reducing magic numbers.

These changes improve code maintainability, readability, and robustness.

JadeCheah committed:
Refactor handleUserInput() for better readability

The handleUserInput() method contained too much logic in one place, making it harder to read and maintain. Extracting helper methods improves readability and makes the code more modular.

Why?
 - Extracted displayUserInput(), processUserCommand(), and exitApplication() from handleUserInput().
 - Reduced deep nesting and improved method separation.
 - Made the main user input handling flow more concise.

JadeCheah committed:
Add TASK_INDEX_OFFSET to remove magic numbers

JadeCheah committed:
Refactor AddCommand: Improve readability, error handling, and JavaDocs
- Extracted logic from `execute()` into smaller helper methods for better readability.
- Added JavaDoc comments to all methods for better documentation.
- Improved error handling in `createTask()` to catch invalid task types.
- Made `TaskType` enum private since it is only used within `AddCommand`.
- Refactored `createDeadline()` and `createEvent()` to handle missing date formats more explicitly.
- Ensured exception chaining for better debugging in `InvalidCommandException`.

These changes improve code readability, maintainability, and error handling consistency.

JadeCheah committed:
Migrate task format validation logic to Parser

JadeCheah committed:
Change parseCommand to better code quality

JadeCheah committed:
Merge pull request #1 from JadeCheah/branch-A-Assertions

Add assertions to enforce key assumptions

JadeCheah committed:
Merge branch-A-CodeQuality into master

JadeCheah committed:
Add assertions

JadeCheah committed:
Add basic tagging features, can add, cannot delete yet, and is buggy

JadeCheah committed:
Do tagging feature

JadeCheah committed:
Merge branch 'branch-C-Tagging' into master

Complete tagging system

Now the chatbot supports:
- adding tags when creating a new task
- adding tags (one at a time) to an existing task
- untagging tags from an existing task

JadeCheah committed:
Add GUI

JadeCheah committed:
Add more GUI changes

JadeCheah committed:
Merge branch 'branch-A-BetterGui' into master

JadeCheah committed:
Add TaskTest and modify AddCommandTest and TaskListTest

JadeCheah committed:
Merge branch 'branch-A-MoreTesting' into master

-Add TaskTest and Modify the other two tests
-Do some touchup on the error handling of FindCommand and TagCommand

JadeCheah committed:
Modify README file and add Ui.png

JadeCheah committed:
Modify README.md

JadeCheah committed:
Modify Readme

JadeCheah committed:
Modify JavaDoc for Launcher main method

JadeCheah committed:
Enable Assertions

JadeCheah committed:
update root README.md

JadeCheah committed:
Point to correct path UI.png
